### PR TITLE
matching: add support for matching on HTTP trailers

### PR DESCRIPTION
**Pull request by snowp**
Adds support for matching on HTTP request and response trailers.

Refactors the matching classes into a Http::Matching namespace to de-clutter the FM.

Fixes a bug around skipping filter iteration where we would not
skip encoding trailers even if the skip_filter_ flag was set.

Signed-off-by: Snow Pettersen <snowp@lyft.com>

Risk Level: Medium, FM changes that should only touch matching flows
Testing: UTs
Docs Changes: Proto docs
Release Notes: n/a

### CODE REVIEW

File: api/envoy/type/matcher/v3/http_inputs.proto [hunk 1]
```diff
@@ -24,6 +24,16 @@ message HttpRequestHeaderMatchInput {
...
 }
 
+// Match input indicates that matching should be done on a specific request trailer.
+// The resulting input string will be all headers for the given key joined by a comma,
+// e.g. if the request contains two 'foo' headers with value 'bar' and 'baz', the input
+// string will be 'bar,baz'.
+// [#comment:TODO(snowp): Link to unified matching docs.]
+message HttpRequestTrailerMatchInput {
+  // The request trailer to match on.
+  string header_name = 1;
```
Comment (lizan) [inline api/envoy/type/matcher/v3/http_inputs.proto:34]: add header name PGV constraint? (same below)

**Comment by snowp** [comment]
This is ready for review now

File: api/envoy/type/matcher/v3/http_inputs.proto [hunk 2]
```diff
@@ -18,7 +19,20 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 // [#comment:TODO(snowp): Link to unified matching docs.]
 message HttpRequestHeaderMatchInput {
   // The request header to match on.
-  string header_name = 1;
+
```
Comment (mattklein123) [inline api/envoy/type/matcher/v3/http_inputs.proto:22]: del newline

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with small doc comment, thanks.

/wait

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**